### PR TITLE
feat: expand exploreDirectory algorithm

### PR DESCRIPTION
**Pull request by JKRhb**
This PR contains a rough first draft for expanding the `exploreDirectory` algorithm.

***
<a href="https://pr-preview.s3.amazonaws.com/JKRhb/wot-scripting-api/pull/545.html" title="Last updated on Jul 8, 2024, 9:03 AM UTC (dc81bc5)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/wot-scripting-api/545/352fe0c...JKRhb:dc81bc5.html" title="Last updated on Jul 8, 2024, 9:03 AM UTC (dc81bc5)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -790,33 +790,36 @@ <h3>Validating an ExposedThingInit</h3>
...
+          Let |consumedTddThing| be the result of
+          <a href="#creating-consumedthing">creating a <dfn>ConsumedThing</dfn></a>
+          from |directoryTd|.
+          If that fails, resolve |promise| with {{SyntaxError}} and stop.
+        </li>
+        <li>
+          Run the <a>discovery process</a> given |discovery| by
+          either invoking one of the actions associated with the Search API
+          – if the {{ThingDiscoveryProcess/[[filter]]}} argument is not
+          `undefined` or `null` – or reading the `things` property.
```
Comment (zolkis) [inline index.html:819]: This should be expanded into more concise sub-steps.

**Review by zolkis — CHANGES_REQUESTED** [review]

File: index.html (hunk 1 above)
Comment (JKRhb) [reply index.html]: Started the split into sub-steps, however, there is still some more refinement needed.

**Comment by zolkis** [comment]
The algorithm drafted here looks more like an internal implementation detail. It is hard to standardize it in order to expose these details to scripts. Or, in other words, it's hard to distillate simple APIs that cover the programmatic side of discovery, without going too much into protocol details -- since then a script could just use the network/REST API.

I am thinking if we could make examples on how to fetch TD directories using the Discovery spec / network API, as depicted [here](https://www.w3.org/TR/wot-discovery/#discovery-overview). Those will help figuring out if we can distillate a good programmatic API on top of that.

Currently the WoT Discovery introduction mechanisms are definitely encapsulated in the Scripting implementation.

Moreover, the Authentication layer and the Exploration layer, too, are encapsulated. The Scripting API presents another (programmatic) exploration API that connects to the middleman (the Scripting implementation), provisioned as an endpoint in the given WoT solution, and maybe represented (in the future) as an internal slot to the implementation (the WoT API object). True, we are missing a standard provisioning API, but more about this later.

Since currently we are not co-hosting multiple solutions in a single WoT runtime (Scripting implementation, i.e. WoT API object), this works quite well, separating the complexities of the underlying WoT Discovery spec.

We can specify a separate API for implementing the details of the WoT Discovery spec, and that should be in a different conformance class, with its own Security section, same level as WoT Provisioning. In my view, these belong to the next level down in the stack. If we were to standardize these (Provisioning, Discovery introduction + exploration), I would support it, but eventually in a separate deliverable (spec, impl + tests) in the WG. 

If we were to co-host and support multiple WoT/IoT solutions with a single runtime and WoT object, then it needs to be transparent from identification/authentication point of view as well, therefore needing more low level configuration functions -- therefore going one level down as mentioned above. 

Another way to do that is to expose multiple WoT objects for multiple solutions, i.e. an additional API to create (async) a WoT API object, with the given provisioning/initialization routines, including discovery introduction and exploration, resulting in a configured discovery "middleman" internal slot. Once that is done, we obtain the WoT object which exposes the current API functionality.

I can draft an experimental API for this in a separate issue.

**Comment by JKRhb** [comment]
Thank you, @zolkis, for your comment and the proposal in #558! I think we can probably close this PR then.

Should we maybe also close #535 since the discussion is going in a different direction now?

**Pull request closed without merging**